src/main/java/org/opensearch/geospatial/ip2geo/action/DeleteDatasourceTransportAction.java

```diff
@@ -0,0 +1,132 @@
+     * @param lockService the lock service
+     * @param ingestService the ingest service
+     * @param datasourceFacade the datasource facade
+     */
+    @Inject
+    public DeleteDatasourceTransportAction(
+        final TransportService transportService,
+        final ActionFilters actionFilters,
+        final Ip2GeoLockService lockService,
+        final IngestService ingestService,
+        final DatasourceFacade datasourceFacade
+    ) {
+        super(DeleteDatasourceAction.NAME, transportService, actionFilters, DeleteDatasourceRequest::new);
+        this.lockService = lockService;
+        this.ingestService = ingestService;
+        this.datasourceFacade = datasourceFacade;
+    }
+
+    /**
+     * We delete datasource regardless of its state as long as we can acquire a lock
+     *
+     * @param task the task
+     * @param request the request
+     * @param listener the listener
+     */
+    @Override
+    protected void doExecute(final Task task, final DeleteDatasourceRequest request, final ActionListener<AcknowledgedResponse> listener) {
+        lockService.acquireLock(request.getName(), LOCK_DURATION_IN_SECONDS, ActionListener.wrap(lock -> {
+            if (lock == null) {
+                listener.onFailure(new OpenSearchException("another processor is holing a lock on the resource. Try again later"));
+                return;
+            }
+            try {
+                deleteDatasource(request.getName());
+                listener.onResponse(new AcknowledgedResponse(true));
+            } catch (Exception e) {
+                listener.onFailure(e);
+            } finally {
+                lockService.releaseLock(
+                    lock,
+                    ActionListener.wrap(
+                        released -> { log.info("Released lock for datasource[{}]", request.getName()); },
+                        exception -> { log.error("Failed to release the lock", exception); }
+                    )
+                );
+            }
+        }, exception -> { listener.onFailure(exception); }));
+    }
+
+    private void deleteDatasource(final String datasourceName) throws IOException {
+        Datasource datasource = datasourceFacade.getDatasource(datasourceName);
+        if (datasource == null) {
+            new ResourceNotFoundException("no such datasource exist");
+        }
+
+        setDatasourceStateAsDeleting(datasource);
+        datasourceFacade.deleteDatasource(datasource);
+    }
+
+    private void setDatasourceStateAsDeleting(final Datasource datasource) throws IOException {
+        if (isSafeToDelete(datasource) == false) {
+            new OpenSearchException("datasource is being used by one of processors");
+        }
+
+        DatasourceState previousState = datasource.getState();
+        datasource.setState(DatasourceState.DELETING);
+        datasourceFacade.updateDatasource(datasource);
+
+        // Check again as processor might just have been created.
```

Comment:
Is there not a write lock that can be grabbed to delete the datasource?

Reply:
Haven't implemented lock between create ip2geo processor and delete datasource.
Appreciate if you could share some idea on the write lock. If write lock involves interacting with index, it cannot be used during ip2geo processor creation phase because processor creation is a process which changes cluster state and while changing cluster state, we cannot call some or all of OpenSearch API like getting or searching document from an index.

Reply:
hmmm yes Im not sure what the lock would be. I would think there would be something with the JobScheduler LockService.

Reply:
Lock with job scheduler utilize opensearch index which cannot be used in processor creation time.

Reply:
Yes. The thing is, we didn't want to add any custom metadata during design review meeting as it can cause a headache in the future due to many restriction on maintaining backward compatibility.

Reply:
Got it, might be something to consider in the future if this becomes a problem.